featurebyte/session/spark.py

```diff
@@ -40,47 +50,56 @@ class SparkSession(BaseSparkSession):
     use_http_transport: bool
     use_ssl: bool
     source_type: SourceType = Field(SourceType.SPARK, const=True)
-    database_credential: Optional[Union[KerberosKeytabCredential, AccessTokenCredential]]
+    database_credential: Optional[SparkDatabaseCredential]
 
     def __init__(self, **data: Any) -> None:
         auth = None
         scheme = None
         access_token = None
         kerberos_service_name = None
 
-        # perform authentication initialization prior to super().__init__()
-        # so that we can use kerberos cache to authenticate storage
-        database_credential = data.get("database_credential")
-        if database_credential:
-            if isinstance(database_credential, KerberosKeytabCredential):
-                # create a temporary keytab file to specify the KDC
-                with tempfile.NamedTemporaryFile(mode="wb", suffix=".keytab") as keytab_file:
-                    keytab_file.write(database_credential.keytab)
-                    keytab_file.flush()
-                    cmd = ["kinit", "-kt", keytab_file.name, database_credential.principal]
-                    subprocess.run(cmd, check=True)
+        super().__init__(**data)
+
+        if self.database_credential:
+            if isinstance(self.database_credential, KerberosKeytabCredential):
                 auth = AuthType.KERBEROS
                 kerberos_service_name = "hive"
-            else:
+            elif isinstance(self.database_credential, AccessTokenCredential):
                 auth = AuthType.TOKEN
-
-        super().__init__(**data)
+                access_token = self.database_credential.access_token
+            else:
+                raise NotImplementedError(
+                    f"Unsupported credential type: {self.database_credential.type}"
+                )
 
         # determine transport scheme
         if self.use_http_transport:
             scheme = "https" if self.use_ssl else "http"
 
-        self._connection = HiveConnection(
-            host=self.host,
-            http_path=self.http_path,
-            catalog=self.database_name,
-            database=self.schema_name,
-            port=self.port,
-            access_token=access_token,
-            auth=auth,
-            scheme=scheme,
-            kerberos_service_name=kerberos_service_name,
-        )
+        def _create_connection() -> HiveConnection:
+            return HiveConnection(
+                host=self.host,
+                http_path=self.http_path,
+                catalog=self.database_name,
+                database=self.schema_name,
+                port=self.port,
+                access_token=access_token,
+                auth=auth,
+                scheme=scheme,
+                kerberos_service_name=kerberos_service_name,
+            )
+
+        # create a connection
+        try:
```

Comment:
Refactor this to avoid running kinit on each session creation. Instead only run kinit on failure (e.g. ticket does not exist or expired)